utils/gcp/object_storage.cc

```diff
@@ -0,0 +1,820 @@
+            throw;
+        }
+    }
+}
+
+future<> 
+utils::gcp::storage::client::impl::send_with_retry(const std::string& path, const std::string& scope, body_variant body, std::string_view content_type, rest::httpclient::handler_func f, httpclient::method_type op, key_values headers) {
+    co_await send_with_retry(path, scope, std::move(body), content_type, [f](const seastar::http::reply& rep, seastar::input_stream<char>& in) -> future<> {
+        // ensure these are on our coroutine frame.
+        auto& resp_handler = f;
+        auto result = co_await util::read_entire_stream_contiguous(in);
+        resp_handler(rep, result);
+    }, op, headers);
+}
+
+future<rest::httpclient::result_type> 
+utils::gcp::storage::client::impl::send_with_retry(const std::string& path, const std::string& scope, body_variant body, std::string_view content_type, httpclient::method_type op, key_values headers) {
+    rest::httpclient::result_type res;
+    co_await send_with_retry(path, scope, std::move(body), content_type, [&res](const seastar::http::reply& r, std::string_view body)  {
+        gcp_storage.trace("{}", body);
+        res.reply._status = r._status;
+        res.reply._content = sstring(body);
+        res.reply._headers = r._headers;
+        res.reply._version = r._version;
+    }, op, headers);
+    co_return res;
+}
+
+// Get an upload session for the given object
+// See https://cloud.google.com/storage/docs/resumable-uploads
+// See https://cloud.google.com/storage/docs/performing-resumable-uploads
+future<> utils::gcp::storage::client::object_data_sink::acquire_session() {
+    std::string body;
+    if (!_metadata.IsNull()) {
+        body = rjson::print(_metadata);
+    }
+    auto path = fmt::format("/upload/storage/v1/b/{}/o?uploadType=resumable&name={}"
+        , _bucket
+        , _object_name
+    );
+
+    auto reply = co_await _impl->send_with_retry(path
+        , GCP_OBJECT_SCOPE_READ_WRITE
+        , std::move(body)
+        , APPLICATION_JSON
+        , httpclient::method_type::POST
+    );
+
+    if (reply.result() != status_type::ok) {
+        throw seastar::httpd::unexpected_status_error(reply.result());
+    }
+    auto location = reply.reply._headers[LOCATION];
+    gcp_storage.debug("Upload {}/{} -> session uri {}", _bucket, _object_name, location);
+    auto [host, spath, port, https] = rest::parse_simple_uri(location);
+    _session_path = std::move(spath);
+}
+
+using namespace seastar::http;
+using namespace std::chrono_literals;
+
+static future<> backoff(int n) {
+    return seastar::sleep(n * 300ms);
+}
+
+static const boost::regex range_ex("bytes=(\\d+)-(\\d+)");
+
+static bool parse_response_range(const seastar::http::reply& r, uint64_t& first, uint64_t& last) {
+    auto& res_headers = r._headers;
+    auto i = res_headers.find(RANGE);
+    if (i == res_headers.end()) {
+        return false;
+    }
+
+    boost::smatch m;
+    std::string tmp(i->second);
+    if (!boost::regex_match(tmp, m, range_ex)) {
```

Comment:
Why not `std::regex_match`?

Reply:
std::regex has a number of issues (esp. in our stdlib iirc), so most scylla code uses boost::regex instead.